### ISSUE TITLE
RFC: Standardise options and CLI flags

### ISSUE THREAD

**Issue by Rich-Harris** [rollup/rollup#1479]
Working on the documentation has shown how bizarre and confusing some of the options are — particularly the fact that JavaScript API/config file options differ from the equivalent CLI flags.

I think we should change them so that they're consistent. Full list below, but in summary:

* `entry` would become `input`
* `dest` would become `output`
* `moduleName` would become `name` (which makes a lot more sense anyway, since it's the name of the *bundle*)
* `sourceMap` and `sourceMapFile` would become `sourcemap` and `sourcemapFile`
* `useStrict` would become `strict`
* the `-u`/`--id` flags would become `--amd.id`, which would be joined by `--amd.define`

**CLI**|**Proposed change**|**Option**|**Proposed change**
-----|-----|-----|-----
 | | |`acorn`| 
 |`-u`/`--id`|**`--amd.id`, `--amd.define`**|`amd`| 
 |`--banner`| |`banner`| 
 | | |`cache`| 
 | | |`context`| 
 |`-o`/`--output`| |`dest`|**`output`**
 |`-i`/`--input`| |`entry`|**`input`**
 | | |`exports`| 
 | | |`extend`| 
 | | |`external`| 
 |`--footer`| |`footer`| 
 |`-f`/`--format`| |`format`| 
 |`-g`/`--globals`| |`globals`| 
 | | |`indent`| 
 | | |`interop`| 
 |`--intro`| |`intro`| 
 | | |`legacy`| 
 | | |`moduleContext`| 
 |`-n`/`--name`| |`moduleName`|**`name`**
 | | |`noConflict`| 
 | | |`onwarn`| 
 |`--outro`| |`outro`| 
 | | |`paths`| 
 | | |`plugins`| 
 | | |`preferConst`| 
 | | |`pureExternalModules`| 
 |`-m`/`--sourcemap`| |`sourceMap`|**`sourcemap`**
 | | |`sourceMapFile`|**`sourcemapFile`**
 | | |`targets`| 
 | | |`treeshake`| 
 |`--strict`| |`useStrict`|**`strict`**
 | | |`watch`| 

CC @vijithassar

**Comment by fskreuz** [comment]
`sourceMap` might need to stay as is, since it's actually comprised of two words. I do a quick search every time I have to spell it.

I think `globals` and `external` need to be named better. Something like `globalReferences` and `externalDependencies`. `format` could be `outputFormat` to better indicate it's for output.

**Comment by Rich-Harris** [comment]
Hmm — perhaps `--sourcemap` needs to become `--sourceMap` in that case. I kinda prefer the all-lowercase look, though maybe aesthetics aren't the determining factor here.

There's a separate issue about globals and externals — https://github.com/rollup/rollup/issues/1425. Think we probably need to resolve that before renaming those options!

**Comment by Rich-Harris** [comment]
Working on this currently. One thing that occurred to me is that it might make sense to combine `output` and `format` since both are effectively mandatory (`output` isn't technically mandatory if you're using the CLI and don't care about sourcemaps, but in every other situation it is).

Current approach:

```js
// single output bundle
export default {
  input: 'src/main.js',
  output: 'bundle.js',
  format: 'cjs'
};

// multiple output bundles
export default {
  input: 'src/main.js',
  targets: [
    { output: 'bundle.cjs.js', format: 'cjs' },
    { output: 'bundle.es.js', format: 'es' }
  ]
};
```

It's not totally obvious that `format` describes the intended *output* format. If we did this instead, it would solve that problem, and would unify `output`/`format` and `targets` at the same time:

```js
// single output bundle
export default {
  input: 'src/main.js',
  output: {
    file: 'bundle.js',
    format: 'cjs'
  }
};

// multiple output bundles
export default {
  input: 'src/main.js',
  output: [
    { file: 'bundle.cjs.js', format: 'cjs' },
    { file: 'bundle.es.js', format: 'es' }
  ]
};
```

If there's a major downside here, it's that in order to maintain consistency with the CLI we'd have to do this:

```bash
rollup --input src/main.js --output.file bundle.js --output.format cjs
```

But the existing shorthand (which I have to assume most people are using) needn't change.

```bash
rollup -i src/main.js -o bundle.js -f cjs
```

If this is to change, it should probably all happen in one go (rather than one set of changes in 0.48, then another set later), so if anyone has any strong thoughts about this one way or the other, then say so soon, because I'll probably PR this in the next few hours... 😀

**Comment by Rich-Harris** [comment]
Implemented in 0.48. Upgrading a few of my projects to track down any lingering bugs, then will update the docs
